src/Illuminate/Database/Eloquent/Concerns/HasCasts.php

```diff
@@ -0,0 +1,360 @@
+    }
+
+    /**
+     * Determine if the cast type is a custom date time cast.
+     *
+     * @param  string  $cast
+     * @return bool
+     */
+    protected function isCustomDateTimeCast($cast)
+    {
+        return strncmp($cast, 'date:', 5) === 0 ||
+            strncmp($cast, 'datetime:', 9) === 0;
+    }
+
+    /**
+     * Determine if the cast type is a decimal cast.
+     *
+     * @param  string  $cast
+     * @return bool
+     */
+    protected function isDecimalCast($cast)
+    {
+        return strncmp($cast, 'decimal:', 8) === 0;
+    }
+
+    /**
+     * Determine if the given attribute is a date or date castable.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isDateAttribute($key)
+    {
+        return in_array($key, $this->getDates(), true) ||
+            $this->isDateCastable($key);
+    }
+
+    /**
+     * Is the checked value a custom Cast.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isCustomCastable($key)
+    {
+        if ($cast = $this->getCast($key)) {
+            return is_subclass_of($cast, Castable::class);
+        }
+
+        return false;
+    }
+
+    /**
+     * Determine whether a value is Date / DateTime castable for inbound manipulation.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isDateCastable($key)
+    {
+        return $this->hasCast($key, ['date', 'datetime']);
+    }
+
+    /**
+     * Determine whether a value is JSON castable for inbound manipulation.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isJsonCastable($key)
+    {
+        return $this->hasCast($key, ['array', 'json', 'object', 'collection']);
+    }
+
+    /**
+     * Getting the execution result from a user Cast object.
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return mixed
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function fromCustomCastable($key, $value = null)
+    {
+        return $this
+            ->normalizeHandlerToCallable($key)
+            ->get($value);
+    }
+
+    /**
+     * Converting a value by custom Cast.
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return mixed
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function toCustomCastable($key, $value = null)
+    {
+        return $this
+            ->normalizeHandlerToCallable($key)
+            ->set($value);
+    }
+
+    /**
+     * Getting a custom cast instance.
+     *
+     * @param  string  $key
+     * @return \Illuminate\Contracts\Database\Eloquent\Castable
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function normalizeHandlerToCallable($key)
+    {
+        if (! array_key_exists($key, $this->casts_instances)) {
```

Comment:
I don't see any particular advantage or specificity for using that method?
```suggestion
        if (! isset($this->casts_instances[$key])) {
```

Reply:
In my opinion there isn’t much difference, but using the array_key_exists method seems to me visually more pleasant.

Reply:
I object:
- isset is much shorter to read and to understand, also it feels more natural to write isset $castsInstances[$key]) then array_key_exists which, despite its name starting with array, requires they key as the first and the array as the 2nd arg;
- it gives the impression that `null`-ness of values is relevant (which is the big difference between them both), but this impression is wrong => null values are not relevant
- microbench wise, isset is faster

There is simple no reason here to use array_key_exists over isset

Reply:
I do not completely agree. `array_key_exists` is easier to read and the code looks cleaner:
```php
if (! array_key_exists($key, $this->castsInstances)) {
if (! isset($this->castsInstances[$key])) {
```

Also in this tray there are other places where `array_key_exists` is used:

1. [L151](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L151)
2. [L176](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L176)
3. [L321](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L321)
4. [L911](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L911)
5. [L1120](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L1120)
6. [L1165](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L1165)

Reply:
I checked in the test for 100,000 iterations. I agree.